fenics_ice/solver.py

```diff
@@ -492,14 +496,34 @@ def sliding_law(self, alpha, U):
         if sl == 'linear':
             B2 = C
 
-        elif sl == 'budd':
-            N = (1-fl_ex)*(H*rhoi*g + ufl.Min(bed, 0.0)*rhow*g)
+        elif ((sl == 'budd') or (sl == 'corn')):
             U_mag = sqrt(U[0]**2 + U[1]**2 + vel_rp**2)
-            # Need to catch N <= 0.0 here, as it's raised to
-            # 1/3 (forward) and -2/3 (adjoint)
-            N_term = ufl.conditional(N > 0.0, N ** (1.0/3.0), 0)
-            B2 = (1-fl_ex)*(C * N_term * U_mag**(-2.0/3.0))
+            N = (1-fl_ex)*(H*rhoi*g + ufl.Min(bed, 0.0)*rhow*g)
 
+            if sl == 'budd':
+                # Need to catch N <= 0.0 here, as it's raised to
+                # 1/3 (forward) and -2/3 (adjoint)
+                N_term = ufl.conditional(N > 0.0, N ** (1.0/3.0), 0)
+                B2 = (1-fl_ex)*(C * N_term * U_mag**(-2.0/3.0))
+
+            elif sl == 'corn':
+                # see eq 11 of Asay Davis et al, Experimental design for 
+                # three interrelated marine ice sheet and ocean model intercomparison projects ... 
+                # Geosci. Model Dev., 9, 2471–2497
+
+                # Need to catch N <= 0.0 here, as it's raised to
+                # 1/3 (forward) and -2/3 (adjoint)
+                LocalProjectionSolver(H, self.H_DG).solve() 
+                LocalProjectionSolver(U_mag, self.Umag_DG).solve()
+                # to come out if not used
+                N_DG = (1-fl_ex)*(self.H_DG*rhoi*g + ufl.Min(self.bed_DG, 0.0)*rhow*g)
+                N_term = ufl.conditional(N > 0.0, N, 0)
+                # to come out if not used
+                N_DG_term = ufl.conditional(N_DG > 0.0, N_DG, 0)
+#                denom_term = (C**3 * self.Umag_DG + (0.5 * N_DG_term)**3)**(1.0/3.0)
+                denom_term = (C**3 * U_mag + (0.5 * N_term)**3)**(1.0/3.0)
+                B2 = (1-fl_ex)*(C * 0.5 * N_term * U_mag**(-2.0/3.0) / denom_term)
```

Comment:
Potential divide-by-zero / small value? Maybe this is more likely in the P0 case?

Reply:
with N nonnegative, C>0 and vel_rp>0 denom_term will be positive. If C==0 there could be a div/0. im not sure if it's best to address that here or elsewhere. im happy to leave it now.